File: cpp/102_Binary Tree Level Order Traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
	vector<vector<int>> levelOrder(TreeNode* root) {
		vector<vector<int>> res;
		if (!root)
			return res;

		vector<TreeNode*> todo;
		todo.push_back(root);
		while (!todo.empty()) {
			vector<TreeNode*> next;
			vector<int> ans;
			for (auto node : todo) {
				ans.push_back(node->val);
				if (node->left)
					next.push_back(node->left);
				if (node->right)
					next.push_back(node->right);
			}
			todo = next;
			res.push_back(ans);
		}
		return res;
	}
};

// using queue
class Solution {
public:
	vector<vector<int>> levelOrder(TreeNode* root) {
		vector<vector<int>> res;
		if (!root)
			return res;
		queue<TreeNode*> todo;
		TreeNode* curr;
		todo.push_back(root);
		while (!todo.empty()) {
			int n = todo.size()
			vector<int> ans;
			for (int i = 0; i < n; ++i) {
				curr = todo.front();
				todo.pop();
				ans.push_back(curr->val);
				if (curr->left)
					todo.push_back(curr->left);
				if (curr->right)
					todo.push_back(curr->right);
			}
			res.push_back(ans);
		}
		return res;
	}
};


// DFS recursion: shared variable res
class Solution {
public:
	vector<vector<int>> res;
	void recursion(TreeNode* node, int depth) {
		if (!root)
			return;
		if (depth >= res.size())
			res.push(vector<int>());
		res[depth].push_back(node->val);
		recursion(node->left, depth + 1);
		recursion(node->right, depth + 1);
	}
	vector<vector<int>> levelOrder(TreeNode* root) {
		if (!root)
			return res;
		recursion(root, 0);
		return res;
	}
};

File: java/102.binary-tree-level-order-traversal.java
import java.awt.List;
import java.util.ArrayList;
import java.util.Deque;

import javax.swing.tree.TreeNode;

/*
 * @lc app=leetcode id=102 lang=java
 *
 * [102] Binary Tree Level Order Traversal
 */

// @lc code=start
/**
 * Definition for a binary tree node. public class TreeNode { int val; TreeNode
 * left; TreeNode right; TreeNode(int x) { val = x; } }
 */
class Solution {
    public List<List<Integer>> levelOrder(TreeNode root) {
        List<List<Integer>> res = new ArrayList<>();
        Deque<TreeNode> stack = new LinkedList<>();
        stack.push(root);
        TreeNode curr;
        while (!stack.isEmpty()) {
            int size = stack.size();
            List<Integer> level = new ArrayList<>();
            for (int i = 0; i < size; i++) {
                curr = stack.removeFirst();
                if (curr != null) {
                    level.add(curr.val);
                    if (curr.left != null)
                        stack.addLast(curr.left);
                    if (curr.right != null)
                        stack.addLast(curr.right);
                }
            }
            if (level.size() > 0)
                res.add(level);
        }
        return res;
    }

    // recursion
    List<List<Integer>> res = new ArrayList<>();

    public List<List<Integer>> levelOrder(TreeNode root) {
        if (root == null)
            return res;
        helper(root, 0);
        return res;
    }

    private void helper(TreeNode node, int level) {
        if (node == null)
            return;
        if (res.size() == level) {
            res.add(new ArrayList<>());
        }
        res.get(level).add(node.val);
        helper(node.left, level + 1);
        helper(node.right, level + 1);
    }
}
// @lc code=end
